proposals/0000-disambiguating-record-updates-using-type-signatures.rst

```diff
@@ -0,0 +1,546 @@
+Disambiguating record updates using type signatures
+===================================================
+
+.. author:: Adam Gundry
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/537>`_.
+.. sectnum::
+.. contents::
+
+This proposal makes a small adjustment to name resolution for record updates
+when ``DisambiguateRecordFields`` is enabled: where multiple types have
+identically-named fields, expression type signatures can be used to disambiguate
+which type is meant.
+
+
+Motivation
+----------
+
+The ``DuplicateRecordFields`` extension makes it possible to define two records
+with the same field name in a single module.  This raises the question of how to
+interpret a record update such as ``r { f = v }`` if there are two fields ``f``
+in scope.  The original design of ``DuplicateRecordFields`` used a limited form
+of type-directed name resolution in this case, however support for this is
+scheduled to be removed following `proposal #366
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0366-no-ambiguous-field-access.rst>`_.
+GHC 9.2 and 9.4 include a warning when such updates are used.  For example, this module: ::
+
+    {-# LANGUAGE DuplicateRecordFields #-}
+
+    data S = MkS { f :: Char }
+    data T = MkT { f :: Int }
+
+    foo :: S -> S
+    foo x = x { f = 'c' }
+
+results in the following warning on GHC 9.2: ::
+
+    M.hs:7:13: warning: [-Wambiguous-fields]
+        The record update x {f = 'c'} with type S is ambiguous.
+        This will not be supported by -XDuplicateRecordFields in future releases of GHC.
+      |
+    9 | foo x = x { f = 'c' }
+      |             ^^^^^^^
+
+The GHC developers would like to remove support for this (mis)feature and turn
+the warning into a fatal error.  However, the question arises as to how such
+updates should be written instead.
```

Comment:
Why is this considered a misfeature? The type of `x` is clearly not ambiguous (it's `S`), so why does the warning say that it is ambiguous?

Reply:
It's ambiguous at name resolution time. See https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0366-no-ambiguous-field-access.rst

Reply:
I see.